### ISSUE TITLE
Section Planes within empty entities containers should be invalid

### ISSUE THREAD

**Issue by DanRathbun** [SketchUp/api-issue-tracker#574]
#### SketchUp Ruby API Issue

Calling [`#vaild?`](https://ruby.sketchup.com/Sketchup/Entity.html#valid%3F-instance_method) on a [`SectionPlane`](https://ruby.sketchup.com/Sketchup/SectionPlane.html) instance in an empty group or component should return **`false`** instead of **`true`**.

The bounds returned by the section plane is equal to it's parent container's bounds (with respect to transformation, height, width, depth, min, max, etc.) **Ie, the section plane's bounds are equally empty**.

Such a section plane will not be displayed as it has no space to be displayed.

As an alternative, ... the [`Model#drawing_element_visible?`](https://ruby.sketchup.com/Sketchup/Model.html#drawing_element_visible%3F-instance_method) method could always return **`false`** for such a section plane. Currently this new method returns **`true`** for empty section situations that are **not actually displayed.**

**Comment by DanRathbun** [comment]
And another weirdness. 

Both the bounds of the empty group (or component instance) and the section plane within it, return **`false`** when **[`bounds.empty?`](https://ruby.sketchup.com/Geom/BoundingBox.html#empty%3F-instance_method)** is called.

**WHY?** Why would anyone want an empty bounding box with no size to return **`false`** for such a method call ?

**Comment by thomthom** [comment]
`Sketchup::Entity#valid?` is the inverse of `Sketchup::Entity#deleted?` (bad name, and redundant). We can't change the semantic of that now however.

The ultimate goal here is to determine is the section cut is slicing anything? Or that it's visible?

**Comment by DanRathbun** [comment]
> `Sketchup::Entity#valid?` is the inverse of `Sketchup::Entity#deleted?` (bad name, and redundant). We can't change the semantic of that now however.

Ah okay. So then this issue probably needs retitling.

But we also have the issue that the empty bounds of such a section plane return `true` for [`#valid?`]() (which is **not** the inverse of a **`deleted?`** method as the [`BoundingBox`](https://ruby.sketchup.com/Geom/BoundingBox.html) class does not have the latter method,) and return the of **`false`** for [`bounds#empty?`](https://ruby.sketchup.com/Geom/BoundingBox.html#empty%3F-instance_method).

> The ultimate goal here is to determine is the section cut is slicing anything? Or that it's visible?

BOTH. And to fix whatever other API weirdness happens when a section plane cuts nothing, ie has no size.

We have the issue of needing to detect when it is not slicing anything.

What do think such a boolean method would be named ? `#working?`, `#slicing?` ...
... or would you rather that `#active?` be modified to return **`false`** in a case like this ?
(Normally a property getter just returns the property **as set**, so I guess I'm leaning away this from this idea.)

 I guess the workaround is ...
```ruby
section_plane.bounds.diagonal == 0.0
```
... perhaps just add this snippet and a note to the docs and forgo the addition of a specialized query method ?

----
I just tested and if such an empty section plane is made active it's a simple no-op.
It makes no difference if the empty section plane is active or not. There are no API errors or warnings.

----

But, we **do have** the bug that [`Model#drawing_element_visible?`](https://ruby.sketchup.com/Sketchup/Model.html#drawing_element_visible%3F-instance_method) does not work correctly for section planes.

Do you want a separate issue for the method bug (so this could be marked a doc issue) ?

I am also thinking here that the definition for a group or component could have a size, but the instance is scaled so small that it will not be visible to the viewer. (This would be operator error in my opinion.)

I'm not sure **how** [`Model#drawing_element_visible?`](https://ruby.sketchup.com/Sketchup/Model.html#drawing_element_visible%3F-instance_method) **would or should** be effected by LOD settings. This might be a separate "visible at current zoom" issue.

----


?

**Comment by thomthom** [comment]
> BOTH. And to fix whatever other API weirdness happens when a section plane cuts nothing, ie has no size.

I guess a method like `section.cutting?` would address the check for whether it's cutting something.

> But, we do have the bug that Model#drawing_element_visible? does not work correctly for section planes.

And adjusting this method would allow us to determine is the section cut is visible in the viewport.

> I'm not sure how Model#drawing_element_visible? would or should be effected by LOD settings.

LOD? "Level Of Detail"? Or did you mean the model rendering options, such as DrawHidden etc? The method will currently take into account DrawHidden settings.

> This might be a separate "visible at current zoom" issue.
In this case do you mean checking if an entity is within the camera frustum? That's another can of worms, which then leads into "what if an object is obfuscated by another?"

**Comment by DanRathbun** [comment]
> LOD? "Level Of Detail"? 

Yes, I meant as you zoom in or out, would the boolean return change to reflect whether the element was (or could be) displayed.

**Comment by bhyoung** [comment]
As discussed at:
https://forums.sketchup.com/t/sectionplane-bounds-bounding-box-corners-are-all-the-same-values/155037

An additional problem with any section plane is that, with the bounding box corners not available, it is not easy for a Tool developer to provide a way to outline the section plane object. The coefficients provided by  the ".plane" plane provide data for an infinite plane. It is desirable to have the bounding box associated with the 4 corners of the section plane.

**Comment by thomthom** [comment]
Is the representation of a section plane always aligned to the local or working axes? A detail to consider as an array of Point3ds might be more appropriate.

**Comment by Eneroth3** [comment]
Since the visual extents of the section plane doesn't drive the bounds of the parent group/component, they aren't technically the bounds of the section plane. I'm also leaning towards a new method returning the corners.

![2021-04-07_13h42_02](https://user-images.githubusercontent.com/5631725/113861023-07230400-97a7-11eb-83de-a424dae47af8.png)

> Is the representation of a section plane always aligned to the local or working axes? A detail to consider as an array of Point3ds might be more appropriate.

Only if the plane is perpendicular to a drawing axis. Planes can be tilted freely in 3d space.
